sycl/doc/extensions/proposed/sycl_ext_oneapi_forward_progress.asciidoc

```diff
@@ -0,0 +1,553 @@
+    sg1-->wi5(Thread);
+    sg1-->wi6(Thread);
+    sg1-->wi7(Thread);
+    sg1-->wi8(Thread);
+    end
+```
+
+[NOTE]
+====
+Blocking with forward progress guarantee delegation ensures that the forward
+progress guarantees provided by at least one thread of execution in a group
+are at least as strong as the forward progress guarantees provided by that
+group.
+====
+
+Safe coordination of work-items via memory requires careful attention to this
+hierarchy, since the ability of work-items to make progress depends on a
+combination of the forward progress guarantees provided by the work-item
+threads of execution and the forward progress guarantees provided by the
+sub-group, work-group and root-group threads of execution. Often, correct
+execution of a kernel requires knowledge of the forward progress guarantees
+of a certain subset of these threads of execution.
+
+To facilitate reasoning about work-item coordination, the forward progress
+guarantees of threads of execution in SYCL can be qualified by a _coordination
+scope_. This extension defines the following coordination scopes:
+
+- All work-item threads of execution spawned directly or indirectly from the
+ND-range kernel's initial root thread of execution are considered part of the
+same root-group scope.
+- All work-item threads of execution spawned directly or indirectly from the
+same work-group thread of execution are considered part of the same work-group
+scope.
+- All work-item threads of execution spawned from the same sub-group thread of
+execution are considered part of the same sub-group scope.
+- Each individual work-item thread of execution is the sole member of its own
+work-item scope.
+
+[NOTE]
+====
+Qualifying a thread of execution's forward progress guarantees with a coordination
+scope can be thought of as a shorthand for combining the forward progress
+guarantees provided by each thread of execution in the hierarchy.
+For example, consider a hypothetical implementation that creates a separate
+`std::thread` (providing concurrent forward progress guarantees) to execute
+each sub-group, but which permits an unbounded number of work-groups executed
+as tasks (providing parallel forward progress guarantees). With work-group
+coordination scope, each sub-group provides concurrent forward progress
+guarantees: this reflects the guarantees associated with each `std::thread`
+upon creation. With root-group coordination scope, each sub-group provides
+only parallel forward progress guarantees: each sub-group's progress guarantees
+are weakened to reflect that the creation of a sub-group depends upon a
+specific work-group (task) being scheduled and making progress.
+====
+
+
+=== Forward progress guarantees
+
+The `forward_progress_guarantee` `enum` is used to represent the three classes
+of forward progress guarantee defined in the {cpp} core language specification.
+
+[source,c++]
+----
+namespace ext::oneapi::experimental::sycl {
+
+enum class forward_progress_guarantee {
```

Comment:
I don't think we should do this.  P2300 also defines a `forward_progress_guarantee` enum, and does not specify values.  I think we should stay aligned with that, at least for now.

I'm also worried that we might have to add additional progress guarantees at some point (either because SYCL needs them, or because C++ introduces them), and updating previously specified values might break user code.

Reply:
Okay that's a good point, having fixed values could complicate adding additional guarantees in the future, especially if they fit in between existing guarantees or if it's no longer a simpler linear hierarchy.